pravega/pravega
# Issue: 5467 updateReaderGroup hangs when changing configuration from subscriber to non-subscriber

Signed-off-by: pbelgundi <prajakta.belgundi@emc.com>

**Change log description**  
updateReaderGroup() API requires handling stream subscriptions when a ReaderGroup Configuration has stream changes and/or when the Config changes Retention.Type from a non-subscriber or a subscriber or vice-versa. 
These multiple combinations were not handled correctly leading to the 'updating' flag in ReaderGroupConfigRecord staying "true" and never changing to "false" when a ReaderGroup config is changed from subscriber to non-subscriber.

**Purpose of the change**  
Fixes #5467

**What the code does**  
1. Added code to check for transition from subscriber or non-subscriber or vice versa and handle stream subscriptions accordingly in StreamMetadataTasks.updateReaderGroup and updateReaderGroupTask.execute()
2. If the old and new config both have RetentionType=NONE (non-subscriber RG) then there is no need to make any changes with subscriber metadata in StreamMetadata. 
3. In all other cases we need to add the ReaderGroup as a subscriber to Streams in the new config and deleteSubscriber() from Streams present in old config but not in new config.

**How to verify it**  
Added additional unit and integration tests to verify all possible config update scenarios from Subscriber to NonSubscriber and vice versa, changes in Stream without change to RetentionType, etc..
All unit and integration tests should pass.
